### DISCUSSION TITLE
Sending a 404 after checking query string parameters in getServerSideProps()

### DISCUSSION THREAD

**Discussion by pugbyte** [vercel/next.js#21572] [Help]
Howdy Next.js community,

I've recently started using Next.js for the first time and I'm really thrilled with it so far (I'm a long-time developer of client-side-rendered React applications, first time getting to do server-side-rendered). Currently I'm working on migrating my site off an old framework and into Next. I'm in a situation where I have to use `getServerSideProps()` instead of `getStaticProps()`, and the particular page in question is pretty heavily reliant on a query string parameter. Only certain values for this query string parameter are considered valid, so the ideal outcome for me is that I can run a validation on the qs param inside `getServerSideProps()` and respond with a 404 if the param is invalid. And doubly ideal would be an elegant way to do it that presents the user with my custom error page page. Here's what I've tried so far:

### First try:
```
export const getServerSideProps = async context => {
  const { query } = context;
  
  ...
  
  if (invalidQsParam) {
    throw new Error("Invalid query string parameter");
  }
  ...
};
```
The code here is really elegant to me, and it does trigger my custom `_error.js` page, but unfortunately this one results in a 500.

### Second try:
```
export const getServerSideProps = async context => {
  const { query } = context;
  
  ...
  
  if (invalidQsParam) {
    context.res.statusCode = 404;
    context.res.end();
    return { props: {} };
  }
  ...
};
```
The code on this one is a bit less elegant to me, it does get me the 404 response I wanted, but unfortunately doesn't trip my custom `404.js` page (or `_error.js` for that matter).

So, in essence I'm looking for a way to "throw" a 404 from inside `getServerSideProps()` if that makes sense. Could anyone offer some help or insight on this one? It would be much appreciated :)

**Comment by sp3c-dod** [comment]
I'm having the same issue with react-router's useSearchParams.get(). Following in case there is a solution that might help me with my similar issue
